src/hippopt/base/optimization_object.py

```diff
@@ -0,0 +1,64 @@
+from hippopt.common import (
+    Any,
+    ClassVar,
+    Type,
+    TypeVar,
+    Union,
+    abc,
+    copy,
+    cs,
+    dataclasses,
+    np,
+)
+
+TOptimizationObject = TypeVar("TOptimizationObject", bound="OptimizationObject")
+StorageType = Union[cs.MX, np.ndarray]
+
+
+@dataclasses.dataclass
+class OptimizationObject(abc.ABC):
+    StorageType: ClassVar[str] = "generic"
+    StorageTypeMetadata: ClassVar[dict[str, Any]] = dict(StorageType=StorageType)
+
+    def get_default_initialization(
```

Comment:
I'd suggest to use an initialization pattern based on the special `__post_init__` method. It could not be straightforward to understand the reasons at first look if you're not familiar with dataclasses. The major benefit of `__post_init__` is that it integrates natively with `dataclasses.InitVar`, that are attributes that are only used during construction and that are not part of the constructed object. It's very useful if the object needs initialization parameters that you don't want to keep.

Ideally, all parameters and variables should be zero-constructible from the combination of their `__post_init__` method and optional `InitVar`s.

Reply:
The point is that from this class point of view, there is no "storage" yet. As far as I understood, the ``__post_init__`` method is called immediately after construction. But at that moment, you might not have all the info you need to construct your storage, and I did not want to force the user to know this at construction time.

I realize now that the name might be misleading. ``initialization`` in this context does not mean "object initialization", but rather "optimization guess", .i.e. a method to call to get the solver warm start. Hence, it is a method to call after the structure has been fully defined. Do you think a different name can help?

Reply:
Assume you want to create a `JointPositions` that is parametric with the degrees of freedoms, i.e. the user somehow has to pass a construction argument that is not relevant to keep. Right now:

```python
class JointPositions(OptimizationObject):
    dofs: int
    storage: StorageType = dataclasses.field(default=None, metadata=Variable.StorageTypeMetadata)

    def get_default_initialization(self, field_name: str) -> npt.NDArray:
        if field_name != "storage":
            raise ValueError
        return np.zeros(self.dofs)

pos = JointPositions(dofs=10)
```

First, I suspect that the logic based on the name might create headaches in case of composition/inheritance.

Second, the `dofs` field has to be stored in the object.

Alternatively, I'd suggest the following alternative pattern:

```python
class JointPositions(OptimizationObject):

    dofs: dataclasses.InitVar[int]
    storage: StorageType = dataclasses.field(default=None, metadata=Variable.StorageTypeMetadata)

    def __post_init__(dofs: int) -> None:
        # Here the user that creates a new class must initialize storage.
        # Either they pass dofs to trigger default construction, or pass an already initialized StorageType object.

        if storage is None:
            self.storage = np.atleast_2d(np.zeros(dofs))        

        # Optionally (but recommended) check the dimensions.
        # This is useful in case of passing a pre-built StorageType object.
        if self.storage.shape != (1, dofs):
            raise ValueError


pos1 = JointPositions(dofs=10)

s = np.zeros(5)
pos2 = JointPositions(dofs=s.size, storage=s)  # Error (this is necessary for casadi/numpy cross-compatibility)

pos3 = JointPositions(dofs=s.size, storage=np.atleast_2d(s))  # ok
```

In my opinion, this approach would also make the logic much compact when there are many fields to initialize:

```python
class LargeComposedVariables(OptimizationObject):

    init_a: dataclasses.InitVar[int]
    init_b: dataclasses.InitVar[list[str]] 

    var1: Variable1 = dataclasses.field(default=None)
    var2: Variable2 = dataclasses.field(default=None)
    var3: Variable3 = dataclasses.field(default=None)

    storage: StorageType = dataclasses.field(default=None, metadata=Variable.StorageTypeMetadata)

    def __post_init__(init_a: int, init_b: list[str]) -> None:

        # init vars can be shared
        self.var1 = var1 if self.var1 is not None else Variable1(init_a=init_a)
        self.var2 = var2 if self.var2 is not None else Variable2(init_a=init_a)

        self.var3 = var3 if self.var3 is not None else Variable1(init_b=init_b)

        if storage is None:
            self.storage = np.atleast_2d(np.zeros(dofs))

        if self.storage.shape != (1, dofs):
            raise ValueError
```

> As far as I understood, the __post_init__ method is called immediately after construction. But at that moment, you might not have all the info you need to construct your storage, and I did not want to force the user to know this at construction time.

This is exactly what I wanted to avoid, ideally the `OptimizationObject` might be `frozen=True`. I assume that the object is already properly allocated when it's constructed. This assumption might be helpful to simplify the logic to pass from numpy to casadi automatically when needed.

Reply:
> Assume you want to create a `JointPositions` that is parametric with the degrees of freedoms, i.e. the user somehow has to pass a construction argument that is not relevant to keep. Right now:
> 
> ```python
> class JointPositions(OptimizationObject):
>     dofs: int
>     storage: StorageType = dataclasses.field(default=None, metadata=Variable.StorageTypeMetadata)
> 
>     def get_default_initialization(self, field_name: str) -> npt.NDArray:
>         if field_name != "storage":
>             raise ValueError
>         return np.zeros(self.dofs)
> 
> pos = JointPositions(dofs=10)
> ```
> 
> First, I suspect that the logic based on the name might create headaches in case of composition/inheritance.

Composition is handled here https://github.com/ami-iit/hippopt/blob/67c68b9b59852c7e6765c03344554d87d87eb4c1/src/hippopt/base/optimization_object.py#L38-L50 and tested in https://github.com/ami-iit/hippopt/blob/67c68b9b59852c7e6765c03344554d87d87eb4c1/test/test_base.py#L73-L93


Basically, it detects if a field is a storage or another ``OptimizationObject``. If it is the second case, it calls the same method recursively on the field.


> Second, the `dofs` field has to be stored in the object.

Nope, `dofs` in this case it is not needed since by default it keeps the shape.
https://github.com/ami-iit/hippopt/blob/67c68b9b59852c7e6765c03344554d87d87eb4c1/src/hippopt/base/optimization_object.py#L25

The two initializations can already coexist, as I did in the test.

> This is exactly what I wanted to avoid, ideally the OptimizationObject might be frozen=True. I assume that the object is already properly allocated when it's constructed. This assumption might be helpful to simplify the logic to pass from numpy to casadi automatically when needed.

This is a bit too premature to define in my opinion. I prefer to keep flexibility at this stage.

Reply:
While I still do not agree with this initialization pattern since it defeats one of the fundamental dataclasses features, I don't want to block proceeding forward, it's very early and we can change it later. Let's see how it behaves as soon as complexity increases.